piet-direct2d/src/lib.rs

```diff
@@ -555,20 +559,358 @@ impl<'a> TextLayoutBuilder for D2DTextLayoutBuilder<'a> {
+        let pt = layout.hit_test_point(Point::new(23.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 4);
+        assert_eq!(pt.is_trailing_hit, true);
+        let pt = layout.hit_test_point(Point::new(24.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 5);
+        assert_eq!(pt.is_trailing_hit, false);
+        let pt = layout.hit_test_point(Point::new(25.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 5);
+        assert_eq!(pt.is_trailing_hit, false);
+
+        // outside
+        println!("layout_width: {:?}", layout.width()); // 46.916015625
+
+        let pt = layout.hit_test_point(Point::new(48.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 9); // last text position
+        assert_eq!(pt.is_trailing_hit, true);
+        assert_eq!(pt.is_inside, false);
+
+        let pt = layout.hit_test_point(Point::new(-1.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 0); // first text position
+        assert_eq!(pt.is_trailing_hit, false);
+        assert_eq!(pt.is_inside, false);
+    }
+
+    #[test]
+    fn test_hit_test_point_complex() {
+        let dwrite = directwrite::factory::Factory::new().unwrap();
+
+        // Notes on this input:
+        // 5 code points
+        // 5 utf-16 code units
+        // 10 utf-8 code units (2/1/3/3/1)
+        // 3 graphemes
+        let input = "é\u{0023}\u{FE0F}\u{20E3}1"; // #️⃣
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+        println!("text pos 2 leading: {:?}", layout.hit_test_text_position(2, false)); // 6.275390625
+        println!("text pos 2 trailing: {:?}", layout.hit_test_text_position(2, true)); // 18.0
+
+        let pt = layout.hit_test_point(Point::new(2.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 0);
+        assert_eq!(pt.is_trailing_hit, false);
+        let pt = layout.hit_test_point(Point::new(4.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 0);
+        assert_eq!(pt.is_trailing_hit, true);
+        let pt = layout.hit_test_point(Point::new(7.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 2);
+        assert_eq!(pt.is_trailing_hit, false);
+        let pt = layout.hit_test_point(Point::new(10.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 2);
+        assert_eq!(pt.is_trailing_hit, false);
+        let pt = layout.hit_test_point(Point::new(14.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 2);
+        assert_eq!(pt.is_trailing_hit, true);
+        let pt = layout.hit_test_point(Point::new(18.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 9);
+        assert_eq!(pt.is_trailing_hit, false);
+        let pt = layout.hit_test_point(Point::new(19.0, 0.0));
+        assert_eq!(pt.metrics.text_position, 9);
+        assert_eq!(pt.is_trailing_hit, false);
+    }
+
+    #[test]
```

Comment:
I think this should test an astral plane codepoint.

Reply:
great, I'll find one.